server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -99,18 +111,47 @@ public static KnnVectorQueryBuilder fromXContent(XContentParser parser) {
     private Integer numCands;
     private final List<QueryBuilder> filterQueries = new ArrayList<>();
     private final Float vectorSimilarity;
+    private final QueryVectorBuilder queryVectorBuilder;
+    private final Supplier<float[]> queryVectorSupplier;
 
     public KnnVectorQueryBuilder(String fieldName, float[] queryVector, Integer numCands, Float vectorSimilarity) {
+        this(fieldName, queryVector, null, null, numCands, vectorSimilarity);
+    }
+
+    public KnnVectorQueryBuilder(String fieldName, QueryVectorBuilder queryVectorBuilder, Integer numCands, Float vectorSimilarity) {
+        this(fieldName, null, queryVectorBuilder, null, numCands, vectorSimilarity);
+    }
+
+    public KnnVectorQueryBuilder(String fieldName, Supplier<float[]> queryVectorSupplier, Integer numCands, Float vectorSimilarity) {
+        this(fieldName, null, null, queryVectorSupplier, numCands, vectorSimilarity);
+    }
+
+    public KnnVectorQueryBuilder(
+        String fieldName,
+        float[] queryVector,
+        QueryVectorBuilder queryVectorBuilder,
+        Supplier<float[]> queryVectorSupplier,
+        Integer numCands,
+        Float vectorSimilarity
+    ) {
         if (numCands != null && numCands > NUM_CANDS_LIMIT) {
             throw new IllegalArgumentException("[" + NUM_CANDS_FIELD.getPreferredName() + "] cannot exceed [" + NUM_CANDS_LIMIT + "]");
         }
-        if (queryVector == null) {
-            throw new IllegalArgumentException("[" + QUERY_VECTOR_FIELD.getPreferredName() + "] must be provided");
+        if (queryVectorSupplier == null && queryVector == null && queryVectorBuilder == null) {
```

Comment:
`queryVectorSupplier` is only used internally and represents a transition value when using the query builder and waiting for the response. We shouldn't check nor throw for it.